### ISSUE TITLE
StaticCommand - apply operator

### ISSUE THREAD

**Issue by tomasherceg** [riganti/dotvvm#280]
Currently, the `staticCommand` can call a static method, pass some arguments, and use the return value to update a single property in the viewmodel, e.g. `{staticCommand: Prop1 = SomeMethod(...)}`.
However, there is no way to update multiple properties at once.

We should add an operator (let's call it "apply") to be able to update mutliple properties in the viewmodel. 

For example, the binding would look like this:
`{staticCommand: _this <| SomeMethod(...)}`

The `SomeMethod` can return any object (it can be an anonymous one), e.g.:

```
public object SomeMethod() {
    return new { Prop1 = ..., Prop2 = ... };
}
```

Values of all properties of the returned object will be assigned to the expression on the left side of the operator.

**Comment by exyi** [comment]
This is definitely good idea, we definitely need something like that. But when I look at the syntax for just assigning a new value to ViewModel, it seems a bit strange, especially if I'd not know what exactly does the staticCommand do and why have we designed it this way ;)

Maybe we could just transfer the mutated state in the viewModel. It would just collide a bit with the way it works now, but it is not a big issue in this case IMO. Maybe it could be decorated with some flag:

```C#
[AllowStaticCommand(MutatesViewModel = true)]
public void SomeMethod() {
    Prop1 = ...;
    Prop2 = ...;
}
```

Another possibility would be to implement deconstruction operator, like C#7 or F# has for tuples. It would the look like:

```C#
[AllowStaticCommand]
public (string, SomeVM) SomeMethod() {
    return (..., ...);
}
```

```
{staticCommand: (Prop1, Prop2) = SomeMethod(...)}
```

This has the advantage that you don't rely on the name of the properties in anonymous object.

**Comment by tomasherceg** [comment]
The first solution with MutatesViewModel is a problem because it will be difficult to detect which properties have changed.

I was thinking about decomposition too and it would complicate the syntax when you need to return 5 values which is quite common (e.g. addresses etc.)

I thought about the `<|` operator because it looks much simpler to implement in the framework. But we can use other syntax.
My idea was that the decomposition would result in much more changes in the parser.

**Comment by exyi** [comment]
It is pretty easy to detect changes - just diff the json, we already do that for normal postbacks ;)

And I don't think that implementing decomposition would be too hard, when I look at how much time and energy we have spend just fixing the project references for VS15 or .NET Core, I think I would be able to rest few hours with parser/resolver enhancement ;)

Other issue with MutatesViewModel is that when you want to change a property in other object than you are passing as parameter. That's a big deal because we are again in the state "Impossible To Do" not "Incomfortable To Do", but maybe we could implement both the mutable VM and "powerful assignment" since it should be pretty easy to implement it and can be really convenient.

For similar reason I like the deconstruction assignment more - it allows stuff like `(_root.SomethingInRoot, _this.Prop) = SomeMethod()`.

**Comment by tomasherceg** [comment]
1) Again, there is no viewmodel on staticCommand. You can diff the parameters, but not the viewmodel properties. 

2) The situation where you need to update something in one viewmodel and something in the root is not so common. We can implement the decomposition too, but I don't think it will be used as frequently.

**Comment by djanosik** [comment]
The `<|` operator feels strange to me. Even though it's pretty common in functional languages. I would also prefer the solution involving tuples and deconstruction. It makes clear what props are updated and how. But I agree we can probably implement both solution.

**Comment by darilek** [comment]
Hello, I would also prefer the solution with deconstruction syntax. The shorter version with `<|` operator is not very clear (what exactly static command returns and modifies)

**Comment by tomasherceg** [comment]
I know, but here is a typical scenario - loading up address by company number. The binding would look like this:

`{staticCommand: (CompanyName, Street, Street2, City, ZIP, TaxNumber, IsVATPayer) = LookupCompanyAddress(CompanyNumber)}`

It's way longer than I'd like in comparison to `{staticCommand: _this <| LookupCompanyAddress(CompanyNumber)}`.

I don't like the `<|` syntax, maybe `<==` would be better, but I don't know.

**Comment by tomasherceg** [comment]
Also, there is another point - you are able to implement the `LookupCompanyAddress` method so you can use it not only from static commands, but as a regular method on the server, because it returns the same type. 

If you use decomposition, you'd need to use it also in the viewmodel, or you'd just have another method which returns AddressDTO directly.

```
public class OrderViewModel : DotvvmViewModelBase {
    ...
    public OrderViewModel(CustomerFacade customerFacade) {
        this.customerFacade = customerFacade;
    }

    public override Task PreRender(IDotvvmRequestContext context) {
        if (!Context.IsPostBack) {
            this.Address = LookupCompanyDetails(customerFacade.GetDefaultCompanyNumber());
        }
    }

    [AllowStaticCommand]
    public AddressDTO LookupCompanyAddress(string companyNumber) {
         // this function can be called from both server and client, because it returns the same object
         // no decomposition needed
         return ...;
    }
}
```

**Comment by exyi** [comment]
@tomasherceg You can just assign the address to `Address` property, or is there any reason why would all address properties be inlined in the ViewModel?

```
{staticCommand: Address = SomeMethod()}
```

The method that returns tuple can also be used on server, with C#7 or F# you could even use the same deconstruction syntax.

**Comment by tomasherceg** [comment]
The problem is that `{staticCommand: _this = SomeMethod()}` doesn't work in Knockout very well. 
Also, there may be some properties which you want to transfer in one direction and you don't want to just overwrite the object that is already there. That's exactly a case we are solving now with @quigamdev in one app - there are 6 properties and we need to update 5 of them, so you can either have these 5 properties in a separate class and then you can use the `=` operator. But forcing the users to change the structure of a viewmodel is not a good way.

I see deconstruction very impractical for more than 3 properties and it is especially a pain when there is a risk that there will be additional properties needed.

All these issues can be solved easier when you design the viewmodel correctly so you won't need them. The problem is, that it is very difficult a when you need to do some changes in the page, the last thing you need to do is to change structure of the viewmodel because it can break a lot of things.

**Comment by darilek** [comment]
Would it be possible to support all syntax variants below (I dared to suggest operator `<=`)? 

- `{ staticCommand: (P1, P2) <= SomeMethod() }` - for "type purists"
- `{ staticCommand: Target <= SomeMethod() }` - batch update of all existing properties
- `{ staticCommand: Target = SomeMethod() }` - old way for compatibility

**Comment by tomasherceg** [comment]
Yes, I think we need to do all three options becasue there are situations where you need them. 

* `{ staticCommand: (P1, P2) = SomeMethod() }` - for small and fixed number of arguments (e.g. vectors, geo coordinates)
* `{ staticCommand: Target <== SomeMethod() }` - for patching multiple properties
* `{ staticCommand: Target = SomeMethod() }` - for replacing the value

**Comment by djanosik** [comment]
How about the `<-` operator that is used in F# for assignment?

**Comment by darilek** [comment]
@djanosik `<-` looks promising (for two first options)

**Comment by tomasherceg** [comment]
Yep, I like it more than `<|` or `<==`.

**Comment by tomasherceg** [comment]
The third options is not for backward compatibility, you may need it some times too.

**Comment by Mylan719** [comment]
The more I think about it the more it seems to me that operatot `?=` would work.
By `?=` you are saying "assign existing" because if I understand correctly, what it does is, that it fills properties of target that exist in source. Also this meaning could be extended so that  `?=` operator would fill properties of target, that exist or are non-null in source. however sometimes you want to assign null....

`{ staticCommand: (P1, P2) = SomeMethod() }` - this is just assigment
`{ staticCommand: Target ?= SomeMethod() }` - this one is conditional
`{ staticCommand: Target = SomeMethod() }` - this one just asigment also

**Comment by exyi** [comment]
I think that this issue is basically resolved by the variables in bindings #869 Tuple deconstruction would be also nice, but since a more general (even though more verbose) approach is in place we can call it solved.
